### PR TITLE
WINC-1181: Add pod CPU and Memory metrics

### PR DESCRIPTION
**Pull request by sebsoto**
Adds recording rules which translates the windows exporter's CPU and memory metrics into the time series required for display in the OpenShift console.

### CODE REVIEW

**Comment by sebsoto** [comment]
image

**Comment by mtnbikenc** [comment]
@sebsoto This is for Pods not Nodes, correct?  That distinction might be helpful in the title and commit message.

**Comment by sebsoto** [comment]
Correct, I'll add that to the commit/pr message

**Review by jrvaldes — COMMENTED** [review]
/lgtm

Should this be documented? So that users are aware such metrics are not available.

**Comment by jrvaldes** [comment]
What about backports?

File: bundle/manifests/windows-prometheus-k8s-rules_monitoring.coreos.com_v1_prometheusrule.yaml [hunk 1]
```diff
@@ -39,3 +39,9 @@ spec:
     - expr: |
         windows_cpu_info
       record: node_cpu_info
+    - expr: |
+        sum(rate(windows_container_cpu_usage_seconds_total[5m]) * on(container_id) group_left(namespace, pod, container) kube_pod_container_info{container_id!=""}) by (pod,namespace)
+      record: pod:container_cpu_usage:sum
+    - expr: |
+        label_replace(windows_container_memory_usage_private_working_set_bytes * on(container_id) group_left(namespace, pod, container) kube_pod_container_info{container_id!=""},"container","","","")
+      record: container_memory_working_set_bytes
```
Comment (mansikulkarni96) [inline bundle/manifests/windows-prometheus-k8s-rules_monitoring.coreos.com_v1_prometheusrule.yaml:47]: It maybe out of scope here, but a quick thing to give the number of running pods
- expr: |
          windows_container_available
          record: kubelet_running_pods

File: bundle/manifests/windows-prometheus-k8s-rules_monitoring.coreos.com_v1_prometheusrule.yaml [hunk 2]
```diff
@@ -39,3 +39,9 @@ spec:
     - expr: |
         windows_cpu_info
       record: node_cpu_info
+    - expr: |
+        sum(rate(windows_container_cpu_usage_seconds_total[5m]) * on(container_id) group_left(namespace, pod, container) kube_pod_container_info{container_id!=""}) by (pod,namespace)
```
Comment (mansikulkarni96) [inline bundle/manifests/windows-prometheus-k8s-rules_monitoring.coreos.com_v1_prometheusrule.yaml:43]: I may have to test this out but AFAIK pod label is not available through windows_exporter, an option here to remove ` * on(container_id) ` is to do a label mapping in:
[bundle/manifests/windows-exporter_monitoring.coreos.com_v1_servicemonitor.yam]
````
relabelings:
        - action: replace
          regex: (.*)
          replacement:
          sourceLabels:
            - __meta_kubernetes_endpoint_address_target_name
          targetLabel: instance
        - action: labelmap
          source_labels: [container_id]
          replacement: pod`
```

**Review by mansikulkarni96 — COMMENTED** [review]
Thanks for working on this @sebsoto, PTAL at my comments.

File: bundle/manifests/windows-prometheus-k8s-rules_monitoring.coreos.com_v1_prometheusrule.yaml (hunk 2 above)
Comment (sebsoto) [reply bundle/manifests/windows-prometheus-k8s-rules_monitoring.coreos.com_v1_prometheusrule.yaml:43]: correct Windows exporter does not add the pod label
I am adding it via [this method](https://github.com/prometheus-community/windows_exporter/blob/master/docs/collector.container.md#useful-queries)

File: bundle/manifests/windows-prometheus-k8s-rules_monitoring.coreos.com_v1_prometheusrule.yaml (hunk 1 above)
Comment (sebsoto) [reply bundle/manifests/windows-prometheus-k8s-rules_monitoring.coreos.com_v1_prometheusrule.yaml:47]: @mansikulkarni96 I'd prefer that be done as a followup PR as this work is targetting the graphs

**Comment by mansikulkarni96** [comment]
@sebsoto changes look good here, I missed this in my previous review but wouldn't it be beneficial to add some tests here to ensure we continue to get the pod metrics like we do here: https://github.com/openshift/windows-machine-config-operator/blob/a05678b58eea291c2dccc76c225e775beac90b0d/test/e2e/metrics_test.go#L264

**Comment by mansikulkarni96** [comment]
Thanks for working on this @sebsoto !
/approve

File: test/e2e/storage_test.go [hunk 3]
```diff
@@ -70,6 +70,14 @@ func testStorage(t *testing.T) {
 			}
 		}()
 	}
+	t.Run("pod metrics", func(t *testing.T) {
```
Comment (saifshaikh48) [inline test/e2e/storage_test.go:73]: Would such a test be better suited as part of metrics_test.go? May it'd be best in the long run to split the other metrics tests out of the creation suite

**Review by saifshaikh48 — COMMENTED** [review]
/lgtm

Thanks, finally got to understand the query structure.

File: test/e2e/storage_test.go (hunk 3 above)
Comment (sebsoto) [reply test/e2e/storage_test.go:73]: This is dependent on a workload, for the sake of not increasing the test runtime anymore I felt it best to put it in the storage test. Ideally in the future we use the same workload for both the storage test, and the network tests.

**Comment by sebsoto** [comment]
/override ci/prow/aws-e2e-upgrade
Test passed, must-gather failed unrelated to WMCO

File: test/e2e/storage_test.go [hunk 4]
```diff
@@ -70,6 +70,17 @@ func testStorage(t *testing.T) {
...
+		if skipWorkloadDeletion {
```
Comment (saifshaikh48) [inline test/e2e/storage_test.go:74]: Think it's worth leaving a TODO to remove this condition in 10.16?

**Review by saifshaikh48 — COMMENTED** [review]
/lgtm

**Pull request merged**